Feature non linear time averaged msd

gitsirsha opened the pull request:
<!-- 
Insert issue number that this PR fixes (if any) just after 'Fixes #'.
If this PR does not fix an existing issue, consider opening one or remove 'Fixes #' from the PR description.
-->
Fixes #5028 

Changes made in this Pull Request:
<!-- Describe the changes that this PR makes. If applicable, use the following bullet list. --> 
- **msd.py** script was modified to support time-averaged mean square displacement calculation for dump files with non-linear time spacings.

## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [x] Issue raised/referenced? 
 Yes Issue #5028 
 - [x] Tests updated/added? 
 Not Yet
 - [x] Documentation updated/added?
 Updated short documentation at the top of the msd.py as comments
 - [x] `package/CHANGELOG` file updated?
 Not Yet
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!) 
 Yes - Added

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5066.org.readthedocs.build/en/5066/

<!-- readthedocs-preview mdanalysis end -->

gitsirsha committed:
Inditial addition of code to calculate time-averaged msd for non-linear dump

gitsirsha committed:
updated_documentation_v1_plots_removed

gitsirsha committed:
Added_name_in_the_package/AUTHORS

mrshirts commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -1,3 +1,25 @@
+#   Modified by Sirsha Ganguly
+#
+# Modification:
+# '_conclude_modified()' is the new function added to calculate time-averaged mean squared displacement of non-linear dumps
```
Suggest something more descriptive here - _conclude_nonlinear() instead of _conclude_modified()

mrshirts commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -263,7 +285,17 @@
+        if len(sequence) < 2:
+            return True
+        step = sequence[1] - sequence[0]
+        return all(sequence[i+1] - sequence[i] == step for i in range(len(sequence) - 1))
```
So right now, it's failing the test.  My guess is that it's going down the wrong test path, because it's interpreting whatever data is being passed in by the test as not actually being linear (maybe something like 1.0000 gets interpreted at 0.999999).  A couple of options here:

1. Add a keyword like `nonlinear` that controls the logic, rather than determining the logic from the steps. 
2. Using something like assert_almost_equal so floating point errors don't cause the logic to give the wrong answer (i.e. that they are not equally spaced when they actually are).

mrshirts reviewed (commented):
Added some comments on the code.

IAlibay commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -1,3 +1,25 @@
+######## Documentation
```
It's fine to do this later once code has been settled on, but documents exist as module docstring, or API docstring (see the sections encompassed by """ below).

orbeckst replied:
As Irfan said: The comments will not be in the final PR. The first line of the file is  always
```
# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
```
and then comes the standard comment block. The only history information (eg who modified the file or added something) that we keep inside the code is versionchanged/versionadded markup in the doc strings. Everything else is known to be available through the git history and the information in the PR/issue.

Could you please 
- remove the comments
- copy them to the PR information

This will make it a lot easier to discuss the intention of the PR separate from the implementation.

Information on the algorithm (basically what's important for a user to know) will go into the doc string in a *Notes*  section (we follow the [numpy doc string standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) and see https://userguide.mdanalysis.org/stable/contributing_code.html#guidelines-for-docstrings for MDAnalysis specific guidelines on docs).

IAlibay commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -1,3 +1,25 @@
+######## Documentation
+#
+# Issue: The default '_conclude_simple()' function previously calculated time-averaged mean squared displacement considering linear timespacing between the frames even when a dump file with non-linear timestep gap is provided
```
It would be better if these details were in the PR info instead - this is where a reviewer usually looks for them (and ultimately we will likely merge without this).

IAlibay commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -1,3 +1,25 @@
+# Modification:
+# '_conclude_modified()' is the new function added to calculate time-averaged mean squared displacement of non-linear dumps
+    # Note: This new function is generalized for non-linear dumps.
+    #       Moreover, if you use this function to calculate for linear timedump this gives same result as '_conclude_simple()' which is implemented in the default version.
```
Thanks for checking this, it's really what we need to discuss the most here. Ultimately I think the end target is this PR will be to end up with just one method. Having alternate code paths adds to user difficulty and maintenance costs.

IAlibay replied:
Just taking things as they are now (before we get into the weeds).Could you comment on performance? Do the two methods run at similar speeds with the same inputs?

IAlibay replied:
What happens if someone uses the fft based method instead with non linear times? Does that fail?

IAlibay reviewed (commented):
Here are some initial questions before we get into the weeds. I'm keen for us to end up with a single method that works for both use cases if possible.

There are probably things we can do to make that happen, but I first went to get a good idea of where things are at in this current state.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -263,7 +285,17 @@
 class EinsteinMSD(AnalysisBase):
+
+    @staticmethod
+    def is_linear(sequence): # This function is to check if the timesteps in the array is linear or not! If it's linear default MDA code will carry-out other wise the modified code carries!
```
Instead of comments, write doc strings.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +456,46 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+    
+
+    def _conclude_modified(self):
+        from collections import defaultdict
```
imports at top of file, 
```python
import collections
```
then use `collections.defaultdict`

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +456,46 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+
+    def _conclude_modified(self):
+        from collections import defaultdict
+        print("The Dump file has non-linear time spacing")
```
Eventually, remove all `print` – if you want to show status messages, use the logger.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +456,46 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+        from collections import defaultdict
+        print("The Dump file has non-linear time spacing")
+        
+        dump_times = [ts.time for ts in self._trajectory]
```
This iterates through the whole trajectory and may already be expensive!

Maybe unavoidable but keep it in mind.

orbeckst replied:
You would need to use `self._sliced_trajectory` instead of `self._trajectory`. If someone is running it with `run(start=10, stop=1000, step=5)` then `self._sliced_trajectory` will contain the proper frames that were analyzed in `_single_frame`. However, the time information should already by available in [self.times](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times):
```suggestion
        dump_times = self.times
```
This should improve performance as you don't have to iterate through the trajectory again.

orbeckst replied:
And it's already a numpy array.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +456,46 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+        n_frames = len(dump_times)
+        n_atoms = self.n_particles
+        positions = np.zeros((n_frames, n_atoms, 3))
+        positions = self._position_array.astype(np.float64)
```
This overwrites `positions`. Delete the line above?? Not sure what the intent was here.

orbeckst reviewed (changes requested):
Thank you for the PR. That's a good start for a discussion.

Initially it would really help to have the big picture up front, namely a description at a high level what you want to accomplish and how.

Additionally
- Describe the algorithm that you're implementing.
- How does it differ from the existing one (without FFT)?
- Can you provide a rough performance comparison between "linear" and "nonlinear"? (I understand that applying the algorithm for equally spaced frames to a trajectory with variably spaced frames is incorrect but I want to get a sense of the necessary added computational complexity.)

I added other comments throughout but initially the big picture is more important because we are trying to figure out if there are ways to re-use code: the fewer code is duplicated, the lower the maintenance burden that we incur, and the easier it is to test and maintain correctness.

gitsirsha committed:
addressed_review_comments_mainly_removed_linearity_check

gitsirsha commented:
**Thank you @IAlibay @orbeckst and @mrshirts for your time to review the initial PR. Based on your suggestions I have modified a few things:**

  1. Added a keyword argument `non_linear` instead of a check like the initial commit `is_linear`.
     `non_linear` is set to `False` and only set `True` by the user to call the new method `_conclude_non_linear` .
     This is done to check if the GH Actions CI tests get cleared or not as the code executes to the default route.
     
     **_The new method only executes when `fft=False` and `non_linear = True`_**

  2. Removed the comments for documentation at the top. I will update that in the doc strings later once we are satisfied with the code.

  3. Removed the print statements.
  
  4. `import collections` at top of the file.

  5. deleted `positions = np.zeros((n_frames, n_atoms, 3))` as it was not necessary.
  
**Necessity of a new method** : 

-  The `_conclude_simple` method calculated the time-averaged msd relying on the number of frames a particular dump file has but not relying on at what timestep did each frame got dumped at. So, a better method is needed to calculate "time-averaging" not "frame-averaging".

- Moreover If the new method `_conclude_non_linear` is used on a linearly dumped file it generates same msd values as the default `_conclude_simple` method generates. This is because the new method solely uses the timestamps each frame is at which for a linearly dumped file is nothing but the frame index!.

**Algorithm of `_conclude_non_linear` to calculate time-averaged msd** : 

- It creates a dictionary/hash-map with the keys being the time difference between the frames `Delta_t` and values being a list containing the frame to frame `MSD` values `[msd1, msd2, .....]` for that particular Delta t `Dictionary to collect MSDs: {Δt: [msd1, msd2, ...]}`.

- The reference frame is changed with each iteration and each time a new Key (i.e, Delta t) is found it is appended in the dictionary/hash-map.

-  If a duplicate Delta_t is found it is added to the value (msd list) of that specefic Key in the dictionary/hash-map.

- Lastly in the dictionary/hash-map for each Delta_t the msd values inside the list are averaged.


**Example Performance comparison**:

A coarse grained LAMMPS trajectory with N=60,000 and 21 frames (**non-linearly spaced timesteps**) is run with fft turned off:

It took around **2.9s** with the new method (toggled on using `non_linear = True`).
and took around **1.6s** with the default method.

(This is not including the plotting time in the user interface)

![output_default](https://github.com/user-attachments/assets/b1a2d14e-b4e0-4abd-8a55-5dfcab2aa4ac)
![output_new](https://github.com/user-attachments/assets/347f42a8-6124-4004-bdc4-5e5d4837a52b)

mrshirts commented:
The big picture here is that for polymer MSD's, the results are highly nonlinear in the amount of time (unlike with liquids, where the MSD becomes linear very quickly).  In fact, generally they have a power law dependence.  To calculate MSD at long log times ends up requiring a huge amount of data; you get the short time behavior very accurately, and long time behavior less accurately, but you don't really need the short time behavior more accurately than the long time behavior.   

So LAMMPS has a standard option to output the trajectory files where every N steps, you output steps logarithmically: perhaps with spacing (roughly):  1, 10, 100, 1000.  So for every N steps, you are only outputting 4 points.  Then maybe your simulation is 100/N steps long, so ,you will get 100 data points for lags of 1, 10, 100, 1000, etc, and you get equal accuracy for all lags, at a cost of some precision at short times, but using way less storage space (polymer simulations can have a lot of particles!).   Polymer scientists use this a lot.

The issue is that the standard code can't handle this, and I'm actually not sure if it can be handled with FFTs.  So I'm not totally sure if it makes sense to have a single code path (although that would be desirable).  What one should have is that one gets the same answer for LINEAR data (within machine precision) if it goes through either path, which can be automatically checked, so at least any code duplication is validated.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +427,40 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+    def _conclude_non_linear(self):
+        
+        dump_times = [ts.time for ts in self._trajectory]
+        n_frames = len(dump_times)
```
Just use `self.nframes`
```suggestion
        n_frames = self.n_frames
```

Sorry, that's not properly documented but it is created with [AnalysisBase._prepare_sliced_trajectory](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase._prepare_sliced_trajectory) (I just updated the docs ... #5075 )

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +427,40 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+
+        msd_dict = collections.defaultdict(list) # Dictionary to collect MSDs: {Δt: [msd1, msd2, ...]}
+
+        # Looping over all the frames as if the referenced gets shifted frame to frame
```
There's probably a way to write this all-vs-all distance calculation without loops and just with numpy operations and thus increase performance.

But for a start this is ok while you're making sure that your code produces correct results.

orbeckst replied:
We have functions in lib.distances that can do optimized all-vs-all distance calculations for coordinates.

There's also cdist in scipy.

gitsirsha replied:
We will look into optimizing it once the `MSD.results.msds_by_particle` is as per the documentation. Supposing there will be slight performance hit when this is implemented for the non linear method.

orbeckst replied:
Add a TODO comment as reminder to look into optimizing performance of the code

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +427,40 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+            avg_msds.append(avg_msd)
+        
+        self.results.timeseries = delta_t_values
+        self.results.msds_by_particle = avg_msds
```
Turn the contents of results into numpy arrays. Your code should produce results in the same way as documented in https://docs.mdanalysis.org/stable/documentation_pages/analysis/msd.html#MDAnalysis.analysis.msd.EinsteinMSD .

gitsirsha replied:
`self.results.timeseries` produces desired results the same way. 
`self.results.msds_by_particle` implementation is still in progress, will update shortly with the performance check and results.

orbeckst reviewed (changes requested):
Overall I am ok with having a separate `_conclude_nonlinear`. It seems sufficiently different to warrant its own code path.

The next important step is to add tests that check that your code runs and which can be used as regression tests during further code optimization.

I also left some comments on how to use existing features of AnalysisBase in your code.

Additionally, start paying attention to the linter. Run your file through black for autoformatting.

(We'll also ask you to write more docs and add an entry to CHANGELOG.)

gitsirsha committed:
added results delta t values

gitsirsha commented:
**Current changes:**
-`MSD.results.timeseries` now stores the msd values and `MSD.results.delta_t_values` stores the delta t values for which time average msds are calculated (In previous commit it was stored in `results.timeseries`). This makes sure `MSD.results.timeseries` array stores results which is consistent as per the [documentation](https://docs.mdanalysis.org/stable/documentation_pages/analysis/msd.html#MDAnalysis.analysis.msd.EinsteinMSD.results.timeseries).

-Simple plot can be generated with:
`msd =  MSD.results.timeseries`
`ax.plot(MSD.results.delta_t_values, msd)`

  or simply `ax.plot(MSD.results.delta_t_values, MSD.results.timeseries)`
  
  - No performance hit and deals with sliced trajectory
  
**Future implementation in progress:**
The `MSD.results.msds_by_particle` (a 2D array) as mentioned in [msds_by_particle](https://docs.mdanalysis.org/stable/documentation_pages/analysis/msd.html#MDAnalysis.analysis.msd.EinsteinMSD.results.msds_by_particle) is _"The MSD of each individual particle with respect to lag-time"_, For uniform dumps this is very straightforward with each row of the array being the `lag-time`. Each lag (Δt) corresponds to a fixed number of frames, so we can neatly store MSD per particle, per lag-index.

However In the non-linear (irregular Δt) case: `msds_by_particle` is defined as the time-averaged MSD of each particle computed at actual Δt values (i.e., time differences between all frame pairs), rather than fixed frame lags. This accounts for non-uniform sampling and ensures MSD is evaluated based on physical time rather than frame index.

`msds_by_particle[Δt_index, i] = ⟨|rᵢ(t + Δt) − rᵢ(t)|²⟩_t`

In short for the `_conclude_non_linear()` method `MSD.results.msds_by_particle` will be a 2D array with rows being the numbers of unique delta_t values that can be obtained from `MSD.results.delta_t_values` and columns being the MSD for each particle for that specific delta_t. This is also consistent of the array structure mentioned ensuring easy plotting of particle specific msd over time-lags.


I will push once this implementation is done shortly and update you with the performance. Thank you again for your valuable reviews!

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -301,7 +304,7 @@ class EinsteinMSD(AnalysisBase):
     """
 
-    def __init__(self, u, select="all", msd_type="xyz", fft=True, **kwargs):
+    def __init__(self, u, select="all", msd_type="xyz", fft=True, non_linear = False, **kwargs):
```
Document new argument. Should look something like
```reST
non_linear : bool
    (describe what the option enables)

    .. versionadded:: 2.10.0

```
(Make sure that there's space under the versionadded and *visually check the built docs* that it's formatted correctly – under Checks there's the Read The Docs build that has the built docs)

orbeckst replied:
Also add an entry to CHANGELOG under Enhancements documenting the new features including the new kwarg.

orbeckst replied:
Run black over your code, I am pretty sure that it should not have spaces around `=` in argument lists.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -338,6 +342,7 @@ def __init__(self, u, select="all", msd_type="xyz", fft=True, **kwargs):
         # result
         self.results.msds_by_particle = None
         self.results.timeseries = None
+        self.results.delta_t_values = None
```
add to docs (with `.. versionadded:: 2.10.0`)

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -383,10 +388,12 @@ def _single_frame(self):
 
     def _conclude(self):
-        if self.fft:
+        if self.fft == True:
```
Leave as is, that was how it's normally written in Python.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -383,10 +388,12 @@ def _single_frame(self):
         ]
 
     def _conclude(self):
-        if self.fft:
+        if self.fft == True:
             self._conclude_fft()
-        else:
+        elif self.non_linear != True:
             self._conclude_simple()
+        elif self.non_linear:
+            self._conclude_non_linear()
```
Let's group this logically:
```
if LINEAR
   do either FFT or normal
else
   do non-linear
```

This keeps the logic MUCH clearer and is more robust – you don't have an `else` so with weird settings, conclude will not do anything.

orbeckst reviewed (changes requested):
I have some comments on the code structure.

The biggest thing is to add tests that exercise all your code and check that it produces the results you expect. Have a look at the code coverage. It should be >= 94%.

You'll also need to update the docs and the CHANGELOG.

gitsirsha commented:
> I have some comments on the code structure.
> 
> The biggest thing is to add tests that exercise all your code and check that it produces the results you expect. Have a look at the code coverage. It should be >= 94%.
> 
> You'll also need to update the docs and the CHANGELOG.

Hi! Thanks for the feedback. I'm not very familiar with writing tests in Python yet. Could you please guide me or point me to an example from this repo so I can follow the same pattern?
(For now I am running the code locally with an example `lammpsdump` file every time I update the code to check for results)

I see the `codecov/project` is at 93.5%. Can you please let me know how can it be increased is it just by adding tests?

Happy to update the CHANGELOG and docs right after!

orbeckst commented:
## Testing

Writing test takes some time to learn. We use [pytest](https://docs.pytest.org/en/stable/) as our testing framework so you may have to keep looking up things there.

See https://userguide.mdanalysis.org/stable/testing.html#testing for more information.

For the MSD code specifically, start (and add your tests) in [testsuite/MDAnalysisTests/analysis/test_msd.py](https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/analysis/test_msd.py).

## Coverage 
For coverage look in the status check box: Right now it says **codecov/patch – 19.35% of diff**, which means less than 20% of the lines of codes that you touched are being tested. This number should be >94%. You can click on the status check to see which lines specifically are not covered and you can also look at the [Files changed](https://github.com/MDAnalysis/mdanalysis/pull/5066/files) tab on this PR and see the codecov annotations of missing coverage.

<img width="842" height="447" alt="image" src="https://github.com/user-attachments/assets/b624f73e-169f-439d-bafb-3a142d7acd67" />

orbeckst commented:
You should add your lammpsdump data file to the testsuite (see [develop/testsuite/MDAnalysisTests/datafiles.py](https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/datafiles.py)). Make it as small as possible and bz2 compress it. Ideally, it should be much less than 1 MB in size. Otherwise, produce a small file for testing. It does not need to produce scientifically solid results, just exercise all the code that you're writing in the same way that "real" data would.

orbeckst commented:
I am going to assign myself as PR shepherd but if anyone else wants to take over or be involved, please do!

gitsirsha committed:
initial test added changed changelog and docs

gitsirsha commented:
With the new commit:

- CHANGELOG has been edited
- Docstrings has been edited
- An initial Test has been added for test_msd.py

This initial test currently checks if `results.timeseries` and `results.delta_t_values` are matched with the expected value. (_**NOTE:** `results.msds_by_particle` is yet to be implemented in the original `msd.py` code inside the new method_)

It seems like some of the automated tests are failing although when I did `pytest testsuite/MDAnalysisTests/analysis/test_msd.py` locally it passed. I am a little confused about this.

Ran black after each`.py` codes are edited

gitsirsha committed:
added LAMMPSDUMP_non_linear to __all__

gitsirsha commented:
It seems from the errors that the filename needs to be added in `datafiles.py > __all__`

gitsirsha commented:
> 
> It seems like some of the automated tests are failing although when I did `pytest testsuite/MDAnalysisTests/analysis/test_msd.py` locally it passed. I am a little confused about this.
> 
fixed by adding the test file name to `__all__` in `datafiles.py`. 
`codecov/patch` has increased to 100%.

The linter seems to giving errors for `msd.py` although black was run after edits.

orbeckst commented:
Did you run black with version 24? (Check `black --version` and if necessary pin to 24)

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -298,10 +305,12 @@ class EinsteinMSD(AnalysisBase):
         Number of particles MSD was calculated over.
 
 
-    .. versionadded:: 2.0.0
```
This has to stay. It records when the *class* was originally added.

Instead, you should add a `.. versionchanged:: 2.10.0` with a short explanation. We'll also add `.. versionadded` to the new kwarg, see comment above.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -290,6 +295,8 @@ class EinsteinMSD(AnalysisBase):
     results.msds_by_particle : :class:`numpy.ndarray`
         The MSD of each individual particle with respect to lag-time.
+    results.delta_t_values : :class:`numpy.ndarray`
+        Array of unique Δt (time differences) at which time-averaged MSD values are computed.
```
Indicate when this new results attribute was added so that users understand which version of the code they need
```suggestion
        Array of unique Δt (time differences) at which time-averaged MSD values are
        computed.

        .. versionadded: 2.10.0

```

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -281,6 +282,10 @@ class EinsteinMSD(AnalysisBase):
+    non_linear : bool
+        If ``True``, calculates MSD for trajectory where frames are
+        non-linearly dumped. To use this set `fft=False`.
+        Defaults to ``False``.
```
Indicate when this new kwarg was added so that users understand which version of the code they need in order to use it
```suggestion
        Defaults to ``False``.
       
        .. versionadded:: 2.10.0
       
       
```
(needs 2 newlines below so that formatting works, manually check the rendered docs from the status checks)

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -298,10 +305,12 @@ class EinsteinMSD(AnalysisBase):
 
 
-    .. versionadded:: 2.0.0
+    .. versionadded:: 2.10.0
```
```suggestion
    .. versionadded:: 2.0.0
    .. versionchanged:: 2.10.0
       Added ability to calculate MSD from samples that are not linearly spaced with the 
       new `non_linear` keyword argument.
```

orbeckst replied:
Just FYI:

Btw, if a review added something as a suggestion, you can just directly commit the suggestion or add multiple suggestions to a batch and then commit the batch, all through the GH web interface. This typically saves you time.

(Generally, "suggestion" really means "this needs to be addressed, either you take it directly or you write it yourself". In some cases it is meant as a point for discussion but these cases should be clear from the context.)

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_msd.py:
```diff
@@ -272,3 +268,144 @@ def test_random_walk_u_fft(self, random_walk_u):
+    result_delta_t = msd.results.delta_t_values
+
+    assert result_msd.ndim == 1
+    assert result_msd.shape[0] > 0
```
assert the exact value that's specific for your test data

orbeckst replied:
Actually, check the full shape
```
assert result_msd.shape == (xxx, )
```
or something similar.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_msd.py:
```diff
@@ -272,3 +268,144 @@ def test_random_walk_u_fft(self, random_walk_u):
+    assert result_msd.shape[0] > 0
+
+    assert result_delta_t.ndim == 1
+    assert result_delta_t.shape[0] > 0
```
assert the exact value that's specific for your test data

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_msd.py:
```diff
@@ -272,3 +268,144 @@ def test_random_walk_u_fft(self, random_walk_u):
+    result_msd = msd.results.timeseries
+    result_delta_t = msd.results.delta_t_values
+
+    assert result_msd.ndim == 1
```
If you assert shape itself then you don't need to check ndim.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_msd.py:
```diff
@@ -272,3 +268,144 @@ def test_random_walk_u_fft(self, random_walk_u):
+    )
+
+    np.testing.assert_allclose(result_msd, expected_msd, rtol=1e-5)
+    np.testing.assert_allclose(result_delta_t, expected_delta_t, rtol=1e-5)
```
At the top of the file, 
```python
from numpy.testing import assert_allclose
```
and then use `assert_allclose()`. See how this is typically done in all the other tests and match the style.

orbeckst commented on testsuite/MDAnalysisTests/data/custom_non_linear/test_non_linear.dump:
Can you bz2 compress this file? Should still work.

orbeckst commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -906,6 +907,7 @@
 
 SURFACE_PDB = (_data_ref / "surface.pdb.bz2").as_posix()
 SURFACE_TRR = (_data_ref / "surface.trr").as_posix()
+LAMMPSDUMP_non_linear = (_data_ref / "custom_non_linear/test_non_linear.dump").as_posix()
```
.bz2

orbeckst replied:
Move near other LAMMPS files and keep them together (also in the `__all__` list)

orbeckst replied:
after line ~785 or so

orbeckst commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -389,6 +389,7 @@
     "SURFACE_PDB",  # 111 FCC lattice topology for NSGrid bug #2345
     "SURFACE_TRR",  # full precision coordinates for NSGrid bug #2345
     "DSSP",  # DSSP test suite
+    "LAMMPSDUMP_non_linear"
```
list near other LAMMPS files

orbeckst replied:
and add a short comment explaining what it is

orbeckst commented on package/CHANGELOG:
```diff
@@ -63,6 +63,11 @@ Enhancements
+   mean square displacement in `package/MDAnalysis/analysis/msd.py`,
+   new bool keyword argument `non_linear=True` under `EinsteinMSD`
+   enables execution of this method.
+   (Issue #5028, PR #5066)
```
Focus on the user-visible changes and keep it short
```suggestion
 * Added capability to calculate MSD from frames with irregular (non-linear)
   time spacing in analysis.msd.EinsteinMSD with keyword argument
   `non_linear=True` (Issue #5028, PR #5066)
```

Also, traditionally we add the newest changes under the heading, i.e., move it to directly under "Enhancements"

gitsirsha committed:
ran black 24.10.0

gitsirsha commented:
> Did you run black with version 24? (Check `black --version` and if necessary pin to 24)

I was running black v25. Downgraded and ran black 24.10.0 with `[c90f52e]`

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_msd.py:
```diff
@@ -272,3 +268,144 @@ def test_random_walk_u_fft(self, random_walk_u):
+
+
+def test_msd_non_linear():
+    u = mda.Universe(LAMMPSDUMP_non_linear, format="LAMMPSDUMP")
```
Make the universe a fixture, e.g. `u_nonlinear`. See near top of the file as examples.

orbeckst reviewed (changes requested):
Good progress! There are still some major things to work on:
* Test that your code also works with different start/stop/step
* test that it works with not only xyz but also other combinations

And some minor/easy to do:
* update docs/comments (see inline comments)
* revert black formatting for any files that you didn't touch — _only reformat the files that you've been working on_, please.

gitsirsha commented:
>     * revert black formatting for any files that you didn't touch — _only reformat the files that you've been working on_, please.
Can you please let me know how to revert back the changes for the recent commit locally. When I ran black `v24.10.0` on `msd.py` it said `1 file unchanged` So, I did `black .`, which changed every `.py` file.

Thank you for all the suggestions will get to those shortly!

orbeckst commented:
If you look at https://github.com/MDAnalysis/mdanalysis/pull/5066/files you see that 260(!) files were changed. I would look into reverting the commit that did that.

Alternatively, find the files where you definitely added code (msd.py. test_msd.py, CHANGELOG, AUTHORS, ... ?) and revert everything else. Look at `git revert`, `git reset` (be careful), and also `git restore`. I also saw "See "Reset, restore and revert" in git(1) for the differences between the three commands.".

gitsirsha committed:
Revert "ran black 24.10.0"

This reverts commit c90f52e59b6db649524998146c96f5d89a266947.

gitsirsha committed:
added test for all msd types - added test with start stop step

gitsirsha committed:
ran black 24.4.2

gitsirsha committed:
ran black on datafiles.py

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -299,9 +313,20 @@ class EinsteinMSD(AnalysisBase):
 
 
     .. versionadded:: 2.0.0
+    .. versionadded:: 2.10.0
```
Should be `.. versionchanged:: 2.10.0`

(The "added" means the class/kwarg/attribute was completely added new.)

gitsirsha commented:
I added the tests to calculate msds for all dimensions and using the start/stop/step in separate functions inside a new class. Let me know if that's how it should be or if there's any way to do it in fewer lines of code for `test_msd.py`

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -316,6 +341,9 @@ def __init__(self, u, select="all", msd_type="xyz", fft=True, **kwargs):
             If ``True``, uses a fast FFT based algorithm for computation of
             the MSD. Otherwise, use the simple "windowed" algorithm.
             The tidynamics package is required for `fft=True`.
+        non_linear : bool
```
Looking at this, it seems that all the docs under `__init__` just duplicate the class-level docs. The class docs are shown in [the sphinx docs](https://mdanalysis--5066.org.readthedocs.build/en/5066/documentation_pages/analysis/msd.html#MDAnalysis.analysis.msd.EinsteinMSD) as far as I can tell.

Can you please _remove_ the whole doc string under `__init__` (both your addition and _all_ of the rest), as a clean-up? Thanks.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
The docs near the top have a formatting issue (line break between "with a" and "simple plot") under the [Computing an MSD](https://mdanalysis--5066.org.readthedocs.build/en/5066/documentation_pages/analysis/msd.html#computing-an-msd) section (see image). Can you please also fix this while you're working on this file?

-----

<img width="733" height="254" alt="image" src="https://github.com/user-attachments/assets/7edb02fb-723f-4750-8222-dbe4699c96d3" />

orbeckst reviewed (changes requested):
Looking really good, almost there, just a few minor comments.

I resolved the ones from previous reviews that you had addressed.

orbeckst commented:
The tests look ok to me. The black-formatting is a bit annoying as it takes up so much vertical space but I don't know how to change it and it's not really important so let's leave it.

You're testing all your functionality, so that's great!

gitsirsha commented:
Thank you! I will do it shortly. Yes I also saw the commit suggestion button but was a bit skeptical to use it as I thought I have to edit the changes locally once again. Thanks for the clarification.

orbeckst commented:
When you commit a suggestion, it is added to the branch on GitHub. You can then pull in these changes with `git pull`.

gitsirsha committed:
cleanup - removed docs and edited formatting

gitsirsha commented:
I'm close to implementing `results.msds_by_particle` for now, it's just a null array. Once that's done I'll add the corresponding tests to the new class and functions following the same pattern.

gitsirsha committed:
added results.msds_by_particle for non_linear method

gitsirsha commented:
> **Future implementation in progress:** The `MSD.results.msds_by_particle` (a 2D array) as mentioned in [msds_by_particle](https://docs.mdanalysis.org/stable/documentation_pages/analysis/msd.html#MDAnalysis.analysis.msd.EinsteinMSD.results.msds_by_particle) is _"The MSD of each individual particle with respect to lag-time"_, For uniform dumps this is very straightforward with each row of the array being the `lag-time`. Each lag (Δt) corresponds to a fixed number of frames, so we can neatly store MSD per particle, per lag-index.
> 
> However In the non-linear (irregular Δt) case: `msds_by_particle` is defined as the time-averaged MSD of each particle computed at actual Δt values (i.e., time differences between all frame pairs), rather than fixed frame lags. This accounts for non-uniform sampling and ensures MSD is evaluated based on physical time rather than frame index.
> 
> `msds_by_particle[Δt_index, i] = ⟨|rᵢ(t + Δt) − rᵢ(t)|²⟩_t`
> 
> In short for the `_conclude_non_linear()` method `MSD.results.msds_by_particle` will be a 2D array with rows being the numbers of unique delta_t values that can be obtained from `MSD.results.delta_t_values` and columns being the MSD for each particle for that specific delta_t. This is also consistent of the array structure mentioned ensuring easy plotting of particle specific msd over time-lags.
> 
> I will push once this implementation is done shortly and update you with the performance. Thank you again for your valuable reviews!

In the latest commit -
i. `msds_by_particle` is added in `msd.py` and some tests are added. The results are huge so I though to keep it to xyz only.
ii. code blocks for sorting delta_t keys and making average msd is reduced
<img width="521" height="47" alt="Screenshot 2025-07-28 at 10 10 44 PM" src="https://github.com/user-attachments/assets/534b65de-641a-4afb-9936-040bb1bf8139" />
iii. In the new method code can be conceptually separated in blocks - Initialization block, calculation block, results/averaging block (following correct format as per docs), actually storing the results in `self.results....` block
<img width="813" height="967" alt="Screenshot 2025-07-28 at 10 14 53 PM" src="https://github.com/user-attachments/assets/a7c37ecd-20d5-4d29-ac34-5a55ba1e9b56" />

gitsirsha commented:
Assigning self.results.timeseries = self.results.msds_by_particle.mean(axis=1) (as done in the two default methods) yields same result. This further ensures consistency in shape and how msds_by_particle is computed and propagated through the algorithm.

gitsirsha commented:
The trade off of times between executing both the methods (simple and the non-linear) does not seem that much -

Here is a local performance comparison - 
**test file = CG dump; N=60000; Frames=21**

Default method:
<img width="736" height="763" alt="Screenshot 2025-07-28 at 10 38 15 PM" src="https://github.com/user-attachments/assets/c0b0051e-5033-4a43-836f-52f08610f223" />
Modified method:
<img width="716" height="579" alt="Screenshot 2025-07-28 at 10 38 52 PM" src="https://github.com/user-attachments/assets/df64d179-5f4a-41b4-9a8e-284fc00b795e" />

There's an internal loop in the interface code that loops over the dimensions that is also taking some time.
Both takes about 1s less without the plots. I presume it is handling huge files well.

gitsirsha committed:
minor chancge - versionadded to versionchanged

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +435,48 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+        # For each delta_t, stacked all squared_disp arrays and averaging over axis=0 (time origins)
+        delta_t_values = sorted(msd_dict.keys())
+        avg_msds = [np.mean(msd_dict[dt]) for dt in delta_t_values]
+        msds_by_particle_array = np.zeros((len(delta_t_values), n_atoms))
```
make it dtype=float64 here, then you don't need to cast below

gitsirsha replied:
I removed the casting and it's by default projecting float64 because `positions = self._position_array.astype(np.float64)`. So, I think it's not required all together.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -421,3 +435,48 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+        self.results.delta_t_values = np.array(
+            delta_t_values, dtype=np.float64
+        )
+        self.results.msds_by_particle = np.array(
```
Why not just
```python
self.results.msds_by_particle = msds_by_particle_array
```
I don't think you need to make it an array again.

gitsirsha replied:
Yes it is not required again. Thanks for pointing it out.

orbeckst commented on testsuite/MDAnalysisTests/data/custom_non_linear/test_non_linear.dump.bz2:
Could put it into a directory `analysis/msd`?

The test files are not that well organized but maybe we can start here and put things related to analysis code into specific analysis directories.

orbeckst commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -783,6 +784,9 @@
     _data_ref / "lammps/additional_columns.lammpstrj"
 ).as_posix()
+LAMMPSDUMP_non_linear = (
+    _data_ref / "custom_non_linear/test_non_linear.dump.bz2"
```
change directory

gitsirsha replied:
Seems like one of the test is failing because it can't locate the example dump file but I checked and it is present. Doing pytest is also not throwing in any errors.
<img width="381" height="80" alt="image" src="https://github.com/user-attachments/assets/4844d968-b58c-4297-83f5-a9f9b3294864" />

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -290,6 +297,12 @@ class EinsteinMSD(AnalysisBase):
         The averaged MSD over all the particles with respect to lag-time.
     results.msds_by_particle : :class:`numpy.ndarray`
         The MSD of each individual particle with respect to lag-time.
```
Also describe what `msds_by_particle` contains when `non_linear=True` is set.

Put yourself in the shoes of someone who wants to use the feature. What do they need to know — that should be in the docs!

orbeckst replied:
If there's no difference then you don't have to update.

gitsirsha replied:
I will edit it soon. Thanks!

gitsirsha replied:
I will proceed with editing the documentation once the issue with the failing check is resolved.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -290,6 +297,12 @@ class EinsteinMSD(AnalysisBase):
     results.msds_by_particle : :class:`numpy.ndarray`
         The MSD of each individual particle with respect to lag-time.
+    results.delta_t_values : :class:`numpy.ndarray`
+        Array of unique Δt (time differences) at which time-averaged MSD values are
```
Does this array also exist when `non_linear=False`?

Explain what's in there in the two different cases.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -299,24 +312,20 @@ class EinsteinMSD(AnalysisBase):
+    ):
         if isinstance(u, groups.UpdatingAtomGroup):
             raise TypeError(
                 "UpdatingAtomGroups are not valid for MSD " "computation"
```
join these two strings, please — just as a clean-up (I know that this isn't your code)

orbeckst reviewed (approved):
Great progress! Quick comments inline. 

@mrshirts could you please also have a look (e.g., are the docs correct and sufficient)? Thanks!

(EDIT: I am still requesting changes, I just pressed the wrong button so that's why this looks like an approval — still it's close to being done!)

orbeckst reviewed (changes requested):
Great progress! Quick comments inline. 

@mrshirts could you please also have a look (e.g., are the docs correct and sufficient)? Thanks!

gitsirsha committed:
dtype removal and cleanup misplaced strings

orbeckst commented:
Your test file is not packaged, see **build_testsuite_sdist** in https://github.com/MDAnalysis/mdanalysis/actions/runs/16632261268/job/47064413158?pr=5066 

YOu'll need to add a line in testsuite/pyproject.toml after https://github.com/MDAnalysis/mdanalysis/blob/96f0b67f97d23b11c3ea627ec491efe87b3c900a/testsuite/pyproject.toml#L118 like
```
"data/analysis/msd/*", 
```
so that the files are included.

You can then manually build the dist file – I'd try
```
cd testsuite
python -m build
```
(`mamba install python-build` if needed) or you can try what the CI does (`pipx run build --sdist --outdir dist testsuite`). Look at the output to check that your file is included. You can also install the resulting dist file to check locally.

gitsirsha committed:
Added entry to project.toml

gitsirsha committed:
Attribute definition edit for cases non_linear : bool

gitsirsha commented:
I’ve updated the definitions and am happy to revise them further if a more concise/better version is suggested. Thank you!

gitsirsha committed:
Indentation edit

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -281,15 +281,34 @@ class EinsteinMSD(AnalysisBase):
-        The MSD of each individual particle with respect to lag-time.
+        The MSD of each individual particle with respect to constant lag-time or
+        unique Δt intervals.
+            for `non_linear=False`: a 2D array of shape (n_lagtimes, n_atoms)
+            for `non_linear=True`: a 2D array of shape (n_delta_t_values, n_atoms)
```
The formatting does not look good when rendered, see [MDAnalysis.analysis.msd.EinsteinMSD.results.msds_by_particle](https://mdanalysis--5066.org.readthedocs.build/en/5066/documentation_pages/analysis/msd.html#MDAnalysis.analysis.msd.EinsteinMSD.results.msds_by_particle) in the docs. Please change the formatting so that it makes sense to a reader.

Always look at the rendered docs. You can read more about reST formatting with sphinx in the [sphinx reST primer](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) and accompanying docs.

<img width="648" height="280" alt="image" src="https://github.com/user-attachments/assets/dcd4d528-9f36-4550-958e-09777bab372a" />

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -281,15 +281,34 @@ class EinsteinMSD(AnalysisBase):
+    results.delta_t_values : :class:`numpy.ndarray`
+        Array of unique Δt (time differences) at which time-averaged MSD values are
+        computed (for `non_linear=True`).
+        For `non_linear=False` it is Null array. Time differences are same as the lag-times.
+        (To access: `lagtimes = np.arange(nframes)*timestep`).
```
Let's change this so that it is also correct in `non_linear=False`, i.e., never a zeros array.

The docs [manually compute `lagtimes`](https://mdanalysis--5066.org.readthedocs.build/en/5066/documentation_pages/analysis/msd.html#computing-an-msd), which is error prone. Now that we have a reason to have the lagtimes in the class, let's make them available for everything.

gitsirsha replied:
Can you please point me to the file from where I can edit the example sections in the docs. I believe it is not in the `msd.py`. I can delete the "lagtimes" computing line. Apologies for the delay, and I appreciate your patience on this PR.
<img width="762" height="289" alt="image" src="https://github.com/user-attachments/assets/1e4fd108-3f03-4b2b-841f-0745b62bd95f" />

orbeckst replied:
It's in msd.py https://github.com/MDAnalysis/mdanalysis/blob/91146c581eee22aa6508a4d45cc271cb4d9d1509/package/MDAnalysis/analysis/msd.py#L114

gitsirsha replied:
Thank you! Sorry for my oversight.

orbeckst reviewed (changes requested):
It makes sense to have `results.delta_t_values` to be **correct for any mode**, not just non-linear; as a user I read the comment and immediately ask myself "If it's so easy, why didn't THEY implement it right away... that's lazy!". Could you please make it so that it is always set and then change the documentation so that `delta_t_values` is used in the examples instead of `lagtimes`?

Please also check formatting of docs.

orbeckst commented:
When reading the docs I also realized that there's the big warning note about "unwrapping" trajectories https://mdanalysis--5066.org.readthedocs.build/en/5066/documentation_pages/analysis/msd.html#computing-an-msd but that's outdated. I'll change it in this PR. You might just have to `git pull` before pushing.

orbeckst committed:
update warning in msd docs

- point to nojump transformation (which has been available since 2.5.0)
- add link to gmx trjconv

gitsirsha committed:
Testing doc formatting, WIP

gitsirsha committed:
Merge branch 'develop' into feature-Non-linear-time-averaged-msd

gitsirsha commented:
`results.delta_t_values` is now available for both simple and fft based methods. `dump_times[1] - dump_times[0]` for a linearly dumped file should be nothing but the "actual time difference between two frames", which is a constant!, previously was instructed to define via `timestep=1` in the docs.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -173,7 +179,7 @@
     end_time = 60
     linear_model = linregress(lagtimes[start_index:end_index],
-    				  		  msd[start_index:end_index])
+                                                  msd[start_index:end_index])
```
weird formatting – tabs/spaces??

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -383,20 +405,27 @@ def _single_frame(self):
     def _conclude_simple(self):
         r"""Calculates the MSD via the simple "windowed" algorithm."""
         lagtimes = np.arange(1, self.n_frames)
+        dump_times = self.times
```
Do we need this? Just use `self.times[0]` and `self.times[1]`.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -383,20 +405,27 @@ def _single_frame(self):
             sqdist = np.square(disp).sum(axis=-1)
             self.results.msds_by_particle[lag, :] = np.mean(sqdist, axis=0)
         self.results.timeseries = self.results.msds_by_particle.mean(axis=1)
+        self.results.delta_t_values = np.arange(self.n_frames) * [
+            dump_times[1] - dump_times[0]
+        ]
```
We're calculating the lag times as `delta_t * [0, 1, 2, ...]`, right? So then we should not use a 1-element list of `[delta_t]`:
```suggestion
self.results.delta_t_values = np.arange(self.n_frames) * (self.times[1] - self.times[0])
```

----

Note that we should actually check that the assumption is true that we're making, namely that really all the delta_t are the same, something like
```python
np.isclose(np.diff(self.times).mean(), self.times[1] - self.times[0])
```

The `run()` method can take the [`frames=...` kwarg](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.run) instead of start/stop/step and if that's the case, we should select your new code path for un-even lag times.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -415,9 +444,54 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
                 or set fft=False"""
             )
 
+        dump_times = self.times
```
same comments as above – no need to introduce a variable for only one use

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -415,9 +444,54 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
                 positions[:, n, :]
             )
         self.results.timeseries = self.results.msds_by_particle.mean(axis=1)
+        self.results.delta_t_values = np.arange(self.n_frames) * [
+            dump_times[1] - dump_times[0]
+        ]
```
same comment as above

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -415,9 +444,54 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+
+    def _conclude_non_linear(self):
+
+        dump_times = self.times
```
remove to make code simpler
```suggestion
```
and then use `self.times` below

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -415,9 +444,54 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+        # Looping over all the frames as if the referenced gets shifted frame to frame
+        for i in range(n_frames):
+            for j in range(i + 1, n_frames):
+                delta_t = dump_times[j] - dump_times[i]
```
```suggestion
                delta_t = self.times[j] - self.times[i]
```

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -415,9 +444,54 @@ def _conclude_fft(self):  # with FFT, np.float64 bit prescision required.
+        n_atoms = self.n_particles
+        positions = self._position_array.astype(np.float64)
+
+        msd_dict = collections.defaultdict(
+            list
+        )  # Dictionary to collect MSDs: {Δt: [msd1, msd2, ...]}
```
Put the comment above the line of code (instead of afterwards) so that the code gets formatted on a single line

orbeckst reviewed (changes requested):
Thanks for the updates. I only have minor cosmetic changes.

While reviewing, I also realized that your code can solve a problem that has been inherent in the existing code: The `run()` method can take the [frames=[0, 20, 25, 100, 1001, ...]` kwarg](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.run) to select any frames instead of start/stop/step and if that's the case, the "linear" approach is incorrect (but that case is not being caught). We should select your new code path for un-even frame separations. 

I am not asking for this to be resolved in this PR. If you agree that this is a problem and that your code could solve it, could you raise an issue for it so that it's not forgotten.

gitsirsha committed:
Update self.results.delta_t_values #1

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

gitsirsha committed:
Update self.results.delta_t_values #2

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

gitsirsha committed:
Update self.results.delta_t_values #3

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

gitsirsha committed:
weird formatting edits

gitsirsha committed:
removed dump_times

gitsirsha commented:
Thank you for the review. 

> Thanks for the updates. I only have minor cosmetic changes.
> 
> While reviewing, I also realized that your code can solve a problem that has been inherent in the existing code: The `run()` method can take the [frames=[0, 20, 25, 100, 1001, ...]` kwarg](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.run) to select any frames instead of start/stop/step and if that's the case, the "linear" approach is incorrect (but that case is not being caught). We should select your new code path for un-even frame separations.
> 
> I am not asking for this to be resolved in this PR. If you agree that this is a problem and that your code could solve it, could you raise an issue for it so that it's not forgotten.

That is a very good observation and yes it is wrong to use the default method for such non-linearly spaced kwargs. I will surely raise an issue and I think just diverting the code path to the method for when such kwargs are present would be a good solution.

gitsirsha commented:
A test can also be added and later modifications can be performed to work the issue out.

gitsirsha commented:
I see that the two codecov tests haven't run. Is there any particular reason? Just curious

orbeckst commented:
No sure why/if codecov is having issues.

orbeckst commented:
Thank you for raising the issue.

orbeckst reviewed (approved):
This all looks great to me, thank you for patiently working with me to address all my comments, @gitsirsha !

I am going to merge your contribution.

orbeckst commented:
Congratulations to your first contribution @gitsirsha 🎉 !

gitsirsha commented:
Thank you for your reviews and guidance. I really appreciate your feedback and patience. It was a great learning experience!
